Formula/l/llvm.rb

```diff
@@ -445,8 +493,38 @@ def install
     end
   end
 
+  # We use the extra layer of indirection in `arch` because the FormulaAudit/OnSystemConditionals
+  # doesn't want to let us use `Hardware::CPU.arch` outside of `install` or `post_install` blocks.
+  def write_config_files(macos_version, kernel_version, arch)
+    clang_config_file_dir.mkpath
+
+    drivers = ["clang", "clang++", "clang-cpp"]
+    arches = Set.new([:arm64, :x86_64])
+    arches << arch
+
+    arches.each do |target_arch|
+      target_triple = "#{target_arch}-apple-darwin#{kernel_version}"
+      drivers.each do |driver|
+        (clang_config_file_dir/"#{target_triple}-#{driver}.cfg").atomic_write <<~CONFIG
```

Comment:
If you're using the `clang-cl` patch then you should be able to just do:

```suggestion
        (clang_config_file_dir/"#{target_triple}.cfg").atomic_write <<~CONFIG
```

or at least that's kinda why I did the `clang-cl` patch.

Reply:
Consider also deploying a `#{arch}-apple-darwin.cfg`, which will be be a fallback for unsupported versions, which uses the unversioned `--sysroot=#{MacOS::CLT::PKG_PATH}/SDKs/MacOSX.sdk`. Hopefully nobody needs it but I think it makes sense to do.

Reply:
Right, I forgot about why we had the `clang-cl` patch.

> Consider also deploying a `#{arch}-apple-darwin.cfg`, which will be be a fallback for unsupported versions, which uses the unversioned `--sysroot=#{MacOS::CLT::PKG_PATH}/SDKs/MacOSX.sdk`. Hopefully nobody needs it but I think it makes sense to do.

What about Xcode-only installs? We also don't handle it currently, and this change breaks them, but not really sure what to do about it.

Reply:
It's realistically difficult/impossible to support Xcode without having a `libxcselect` wrapper. Apple don't even manage it without that.

```console
$ echo | /Library/Developer/CommandLineTools/usr/bin/clang -E -Wp,-v -
clang -cc1 version 16.0.0 (clang-1600.0.26.3) default target arm64-apple-darwin23.6.0
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 /Library/Developer/CommandLineTools/usr/lib/clang/16/include
 /Library/Developer/CommandLineTools/usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.

$ echo | xcrun /Library/Developer/CommandLineTools/usr/bin/clang -E -Wp,-v -
clang -cc1 version 16.0.0 (clang-1600.0.26.3) default target arm64-apple-darwin23.6.0
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /Library/Developer/CommandLineTools/usr/lib/clang/16/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
 /Library/Developer/CommandLineTools/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
```

Some people to install the latest Xcode as `/Applications/Xcode-beta.app` (particularly during beta cycles), some install it in `~/Applications` instead, etc.

Even the existing solution here doesn't support that anyway so not sure what you mean by "breaking" it.

Reply:
Well by "breaking" it, I mean that a user with an Xcode-only install at least has a chance of getting a somewhat-sensible `DEFAULT_SYSROOT` which comes from `MacOS.sdk_path_if_needed` if they build from source:

```
❯ sudo mv /Library/Developer/CommandLineTools .
❯ brew ruby -e 'puts MacOS.sdk_path_if_needed'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.sdk
```

Now everyone for whom this worked for when they built `llvm` from source will still get a toolchain that refers to a non-existent path for the sysroot.

Reply:
Seems to work even if `Xcode.app` is moved:
```
❯ sudo mv /Applications/Xcode.app /Applications/Xcode_15.app
❯ brew ruby -e 'puts MacOS.sdk_path_if_needed'
/Applications/Xcode_15.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.sdk
```

Reply:
> Or we just tell Xcode users to use `xcrun` when running the on Xcode-only installs, which should already work fine and already takes precedence over `--sysroot`.

I'll add here that this only works because we use `--sysroot` and not `-isysroot` (which would take precedence over the `SDKROOT` env var that `xcrun` uses).

Reply:
Actually, is that even true? Quick testing shows that Clang prefers to link with the value from `--sysroot`?

Reply:
Works for me:

```console
$ echo | /opt/homebrew/opt/llvm/bin/clang -E -Wp,-v -
clang -cc1 version 19.1.3 based upon LLVM 19.1.3 default target arm64-apple-darwin23.6.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /opt/homebrew/Cellar/llvm/19.1.3/lib/clang/19/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks (framework directory)
End of search list.
# 1 "<stdin>"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 455 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "<stdin>" 2

$ echo | xcrun /opt/homebrew/opt/llvm/bin/clang -E -Wp,-v -
clang -cc1 version 19.1.3 based upon LLVM 19.1.3 default target arm64-apple-darwin23.6.0
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /opt/homebrew/Cellar/llvm/19.1.3/lib/clang/19/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
# 1 "<stdin>"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 455 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "<stdin>" 2

Reply:
Note that the old behaviour is Xcode-only installs had to install LLVM from source which would take hours. In fact, that behaviour is still there - we haven't removed that yet. If you install LLVM without CLT installed then you will get an error and be told to either install the CLT or build from source - it's just that building from source doesn't probably work at the monent.

CLT being installed remains the recommended way to use LLVM. If you use GCC, it is absolutely required and always has been.

Reply:
What's the output if you do:
```
echo > foo.c
clang -v foo.c
```

My guess is that this will still pass the Homebrew path to `ld` in `-syslibroot`, even when running under `xcrun`?

Reply:
Interesting spot - `-syslibroot` seems to always be the opposite of how the include path behaves, which seems very unusual. The include headers always prioritises `-isysroot`/`SDKROOT` but `-syslibroot` prioritises `--sysroot`.

This behaviour technically hasn't changed. Prior to this PR we set `DEFAULT_SYSROOT` which is the same thing as `--sysroot` except its baked into the binary so changing the SDK with `SDKROOT`/`-isysroot` should behave the same way and indeed it does if you try it with `llvm@18`.

So the main regression here is still that Xcode-only installs aren't properly handled yet out of the box for people who have chosen to `brew install --build-from-source llvm` rather than install the CLT.

Reply:
Thanks a lot for the answers! I'll open a separate issue about `-syslibroot`

Reply:
Would also be great to hear more on what people on Xcode-only installs think given the CLT soft requirement has been there for a while. Has `brew install --build-from-source llvm` been a good experience? How often do you upgrade? Is there a reason why building LLVM from source is seen as a better option to installing the CLT?

Reply:
Sorry, cannot answer those questions, I'm acting here because of https://github.com/rust-lang/cc-rs/issues/1278 and https://github.com/rust-lang/rust/pull/131477, and hence it is somewhat important to me that Homebrew gets the details here right, but I might have been completely mistaken in certain regards (again, sorry!) because I haven't used Homebrew myself for a while.

Reply:
Ah good point about missing `*-apple-macosx` - that's definitely worth fixing. I remembered about them initially but completely forgot when this PR was opened.

Reply:
Opened https://github.com/Homebrew/homebrew-core/issues/197277 and https://github.com/Homebrew/homebrew-core/issues/197278